xds/src/main/java/io/grpc/xds/internal/certprovider/ZatarCertificateProvider.java

```diff
@@ -0,0 +1,189 @@
+    final InternalLogId logId = InternalLogId.allocate("ZatarCertificateProvider", details);
+    return new SynchronizationContext(
+        new Thread.UncaughtExceptionHandler() {
+          private boolean panicMode;
+
+          @Override
+          public void uncaughtException(Thread t, Throwable e) {
+            logger.log(
+                Level.SEVERE,
+                "[" + logId + "] Uncaught exception in the SynchronizationContext. Panic!",
+                e);
+            panic(e);
+          }
+
+          void panic(final Throwable t) {
+            if (panicMode) {
+              // Preserve the first panic information
+              return;
+            }
+            panicMode = true;
+            close();
+          }
+        });
+  }
+
+  @Override
+  public void start() {
+    scheduleNextRefreshCertificate(/* delayInSeconds= */0);
+  }
+
+  @Override
+  public void close() {
+    if (scheduledHandle != null) {
+      scheduledHandle.cancel();
+      scheduledHandle = null;
+    }
+    getWatcher().close();
+  }
+
+  private void scheduleNextRefreshCertificate(long delayInSeconds) {
+    if (scheduledHandle != null && scheduledHandle.isPending()) {
+      logger.log(Level.SEVERE, "Pending task found: inconsistent state in scheduledHandle!");
+      scheduledHandle.cancel();
+    }
+    RefreshCertificateTask runnable = new RefreshCertificateTask();
+    scheduledHandle =
+        syncContext.schedule(runnable, delayInSeconds, TimeUnit.SECONDS, scheduledExecutorService);
+  }
+
+  @VisibleForTesting
+  void checkAndReloadCertificates() {
+    try {
+      Path targetPath = Files.readSymbolicLink(directory);
+      if (targetPath.equals(lastModifiedTarget)) {
+        return;
+      }
+      PrivateKey privateKey =
+          CertificateUtils.getPrivateKey(
+              new FileInputStream(new File(targetPath.toFile(), privateKeyFile)));
+      X509Certificate[] certs =
+          CertificateUtils.toX509Certificates(
+              new FileInputStream(new File(targetPath.toFile(), certFile)));
+      X509Certificate[] caCerts =
+          CertificateUtils.toX509Certificates(
+              new FileInputStream(new File(targetPath.toFile(), trustFile)));
+      getWatcher().updateCertificate(privateKey, Arrays.asList(certs));
+      getWatcher().updateTrustedRoots(Arrays.asList(caCerts));
+      lastModifiedTarget = targetPath;
+    } catch (Throwable t) {
+      generateErrorIfCurrentCertExpired(t);
+    } finally {
+      scheduleNextRefreshCertificate(refreshIntervalInSeconds);
+    }
+  }
+
+  private void generateErrorIfCurrentCertExpired(Throwable t) {
+    X509Certificate currentCert = getWatcher().getLastIdentityCert();
+    if (currentCert != null) {
+      long delaySeconds = computeDelaySecondsToCertExpiry(currentCert);
+      if (delaySeconds > GRACE_INTERVAL_IN_SECONDS) {
+        logger.log(Level.FINER, "reload certificate error", t);
+        return;
```

Comment:
This seems generate error if current cert _not_ expired. Maybe my understanding is wrong. Can you explain the logic here?

Reply:
Sure. As we had specified in the design doc the CertificateProvider will generate error (i.e. notify thru watcher's `onError`) *only if* there is no current valid certificate or the last certificate had expired. Here I have provided for some grace period (4 seconds) so I generate the error if the certificate is about to expire in 4 seconds (instead of only after the cert has actually expired).

But I can see why there is confusion. I will remove GRACE_INTERVAL_IN_SECONDS and instead just use the refreshInterval as the grace period since it is an error-worthy event if the cert is about to expire in the next 10 minutes and my last cert refresh attempt failed.